### PR TITLE
Implement George's feedback in PR #10

### PR DESCRIPTION
**Pull request by GeorgeFlerovsky**
## Code formatting to align with the spec

The convention Aiken/Plutus style of a bunch of let/expect statements followed by a monster `and` block makes it very hard to compare the code to the spec.
We'll have a much easier time comparing the spec to our code if we use the following style.
```
pub fn prepend_unsafe(
  key_to_prepend: NodeKey,
  prepended_node_index: Int,
  anchor_node_output_index: Int,
  node_inputs: List<Output>,
  node_outputs: List<Output>,
  node_mint: Dict<AssetName, Int>,
) -> Bool {
  // # Verify the mint:
  // 1. Let key_to_prepend be the key being prepended.
  // 2. The transaction's sole effect on the list is to add key_to_prepend.
  expect key_added(key_to_prepend, node_mint)

  // # Verify the inputs:
  // 3. node_inputs must be a singleton. Let anchor_node_input be its sole node.
  expect list.length(node_inputs) == 1
  expect Some(anchor_node_input) = list.head(node_inputs)
  expect InlineDatum(anchor_node_input_datum) = anchor_node_input.datum
  expect parsed_anchor_node_input_datum: NodeDatum = anchor_node_input_datum

  // ...

  // # Verify immutable data:
  // 10. anchor node input must match anchor node output on address, value,
  //     and datum except for the link field.
  expect anchor_node_input.address == anchor_node_output.address
  expect anchor_node_input.value == anchor_node_output.value
  expect anchor_node_new_datum == parsed_anchor_node_output_datum

  // 11. prepended_node must match anchor_node_output on address.
  expect prepended_node.address == anchor_node_output.address

  // Done.
  True
}
```

After reformatting the code in this way, I was able to detect a few duplicate/redundant conditions in the code, and a few conditions in the code that should be added to the spec. See this PR to the spec:
 https://github.com/Anastasia-Labs/midgard/pull/45

## Clearer names/types

I find the `node_in_outputs` naming convention to be very confusing. My eyes often skip over the `in` and focus on the `outputs`. Moreover, since both `node_outputs` and `node_in_outputs` have the same type `List<Output>`, there's a risk of one of them accidentally being used in place of the other.

Instead, I think we should just rename `node_in_outputs` as `node_inputs`. If we later have a situation where we need to name a variable with type `List<Input>`, we can call it `some_inputs_with_ids`.

## Internal predicates

I moved the old predicate functions to a separate module `linked-list/internal/predicates`. These predicates are considered internal because they assume that they're being called directly from the list handlers and they don't check whether their arguments are valid or sufficient.

These functions should never be called by outside applications.

## External predicates

I added separate predicate functions to the ordered and unordered linked list modules. Each of these predicates are safe to be used by external applications because they check that the list node provided as a witness contains the correct list node token.

These predicates still don't check for the list address because the list handlers already guarantee that list node addresses are correct.

### CODE REVIEW

**Comment by keyan-m** [comment]
Can we aim for a single file? Unless we can push for Aiken to support selective exposure of modules, this seems to be exposing too much details for users.

**Comment by GeorgeFlerovsky** [comment]
> Can we aim for a single file?

You mean a single file for the entire linked list lib?

> Unless we can push for Aiken to support selective exposure of modules

I don't understand what you mean by this.

**Comment by keyan-m** [comment]
> You mean a single file for the entire linked list lib?

Yes precisely. There doesn't seem to be any logical reason to break the file into multiple ones. I believe Evan Czaplicki has a nice talk about this:
https://youtu.be/XpDsk374LDE

It'll also lead to a more consistent interface.
 
> I don't understand what you mean by this.

I'm seeing an `internal` folder which implies its contents are meant to be only accessible by the library and should not be exposed to users. But anything prefixed with a `pub` will be, so I don't think this is optimal.

As for selective module exposure, I am thinking along the lines of the `exposed-modules` stanza of a `.cabal` file. I'll ask around in TxPipe Discord to see if there have ever been any discussions around this.

**Comment by GeorgeFlerovsky** [comment]
> https://youtu.be/XpDsk374LDE

Can you TLDR this video for me plz?

**Comment by keyan-m** [comment]
> They should be separate because each type of list has a specific API that we've defined to be safe for it, and I don't want people to mix and match between the two.

This is the kind of guarantee that should be provided via datatypes. Separation of the modules won't prevent mixing of the two structures.

> Can you summarize this video for me plz?

It's been quite a while since I watched it last. But if I recall correctly, the relevant point it makes is that files shouldn't be broken down into multiple files based on the lengths of their contents, but rather the logic they carry.

**Comment by GeorgeFlerovsky** [comment]
> > They should be separate because each type of list has a specific API that we've defined to be safe for it, and I don't want people to mix and match between the two.
> 
> This is the kind of guarantee that should be provided via datatypes. Separation of the modules won't prevent mixing of the two structures.

Normally, I would agree with you, but datatypes are too expensive in UPLC. Modules are free.

> > Can you summarize this video for me plz?
> 
> It's been quite a while since I watched it last. But if I recall correctly, the relevant point it makes is that files shouldn't be broken down into multiple files based on the lengths of their contents, but rather the logic they carry.

OK, please take a look at the functions in the modules and tell me how you would achieve the same thing with less modules while maintaining separate modules for ordered and unordered lists.

**Comment by GeorgeFlerovsky** [comment]
> I think we should also remove `utils.ak`.

Would it be OK if we move it into the `internal` subfolder and renamed it?

**Comment by keyan-m** [comment]
> Would it be OK if we move it into the `internal` subfolder and renamed it?

If we can't have selective exposure, we should have no `internal` modules (I was referring to both `utils.ak` files by the way).

> OK, please take a look at the functions in the modules and tell me how you would achieve the same thing with a smaller number of semantically-organized modules, while maintaining separate modules for ordered and unordered lists.

Sure :+1: Having two modules might lead to some duplication, however I don't think it's necessarily a bad thing in our case.

I'd still prefer a single module though :sweat_smile:

**Comment by GeorgeFlerovsky** [comment]
> I'd still prefer a single module though 😅

How would you keep things safe with a single module? I want to avoid anyone using the ordered-only methods with unordered lists, and vice versa.

**Comment by GeorgeFlerovsky** [comment]
> If we can't have selective exposure, we should have no internal modules (I was referring to both utils.ak files by the way).

Why not? We can have selective exposure with docs by putting a `//// @hidden` line at the top of the file, and an `internal` subfolder is a strong signal that it should not be imported. If people want to disregard that signal, it's at their own risk.

**Comment by GeorgeFlerovsky** [comment]
I like how Haskell's `containers` package is organized:

- Each data structure has its own module
- If a data structure has more than one variant, those have their own modules.
- Internal modules can be reused by the exported modules, but should not be imported outside of the library.

**Comment by GeorgeFlerovsky** [comment]
@keyan-m Anyway, for now I'll collapse everything down to two modules `linked_list/{ordered,unordered}`.

I'll duplicate the common definitions as needed, so that neither module needs to import from the other.

It's good enough for now and we really need to merge some stuff in. We can have further discussions in early January.

**Comment by GeorgeFlerovsky** [comment]
@keyan-m in my latest commits, I have eliminated all internal modules and eliminated the import of `unordered` functions into `ordered`. Duplicating code as needed.

Note that even though `NodeDatum` is defined in both the ordered and unordered modules, Aiken does distinguish between `ordered.NodeDatum` and `unordered.NodeDatum`.
